File: FloatingWindowEx/app/src/main/java/com/example/floatingwindowex/MainActivity.java
package com.example.floatingwindowex;

import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.content.res.AppCompatResources;

import android.app.ActivityManager;
import android.app.AlertDialog;
import android.content.Context;
import android.content.Intent;
import android.net.Uri;
import android.os.Build;
import android.os.Bundle;
import android.provider.Settings;
import android.text.Editable;
import android.text.TextWatcher;
import android.util.Log;
import android.widget.EditText;
import android.widget.ImageButton;

public class MainActivity extends AppCompatActivity {

    AlertDialog dialog = null;

    ImageButton btnScreenResize;
    EditText edtMessage;


    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        btnScreenResize = findViewById(R.id.btnScreenSize);
        edtMessage = findViewById(R.id.edtMessage);

        btnScreenResize.setImageDrawable(AppCompatResources.getDrawable(this, R.drawable.icon_minimize));


        if (isServiceRunning()) {
            // 이미 플로팅 서비스가 실행중이면 기존 서비스 제거
            stopService(new Intent(this, FloatingWindowService.class));
        }

        edtMessage.setText(Common.currentMessage);
        edtMessage.setSelection(Common.currentMessage.length());
        edtMessage.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {

            }

            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {
                Common.currentMessage = edtMessage.getText().toString();

                Log.i("TAG", "currentMessage:"+Common.currentMessage);
            }

            @Override
            public void afterTextChanged(Editable s) {

            }
        });

        btnScreenResize.setOnClickListener(v -> {
            if (checkOverlayPermission()) {
                startService(new Intent(this, FloatingWindowService.class));
                finish();
            } else {
                requestFloatingWindowPermission();
            }
        });
    }

    private boolean isServiceRunning() {
        ActivityManager am = (ActivityManager) getSystemService(Context.ACTIVITY_SERVICE);

        for (ActivityManager.RunningServiceInfo serviceInfo : am.getRunningServices(Integer.MAX_VALUE)) {
            if (FloatingWindowService.class.getName().equals(serviceInfo.service.getClassName())) {
                return true;
            }
        }
        return false;
    }

    private void requestFloatingWindowPermission() {
        // Overlay Window를 띄우기 위해선 앱 권한 요청이 필요함.
        AlertDialog.Builder builder = new AlertDialog.Builder(this);
        builder.setCancelable(true);
        builder.setTitle("Screen Overlay Permission Required");
        builder.setMessage("Enable 'Display over the App' from Settings");
        builder.setPositiveButton("Open Settings", (dialog, which) -> {
            Intent intent = new Intent(
                    Settings.ACTION_MANAGE_OVERLAY_PERMISSION,
                    Uri.parse("package:" + getPackageName())
            );

            startActivityForResult(intent, RESULT_OK);
        });
        dialog = builder.create();
        dialog.show();
    }

    private boolean checkOverlayPermission() {
        // 안드로이드 M 이상일 경우 시스템 권한 확인
        if (Build.VERSION.SDK_INT > Build.VERSION_CODES.M) {
            return Settings.canDrawOverlays(this);
        } else {
            // M이 아닐 경우 기본적으로 가능
            return true;
        }
    }
}

File: ViewModel2/app/src/main/java/com/puresoftware/viewmodel2/MainActivity.kt
package com.puresoftware.viewmodel2

import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import android.text.Editable
import android.text.TextWatcher
import android.util.Log
import androidx.databinding.DataBindingUtil
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProvider
import com.puresoftware.viewmodel2.databinding.ActivityMainBinding

// https://todaycode.tistory.com/33
class MainActivity : AppCompatActivity() {

    private lateinit var binding: ActivityMainBinding // 이미 viewBinding이 되어버림.
    private lateinit var userViewModel: UserViewModel // 우리가 쓸 ViewModel

    //log
    val TAG: String = MainActivity::class.java.simpleName

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = DataBindingUtil.setContentView(this, R.layout.activity_main) // binding

        //log
        Log.i(TAG, "Oncreated")

        // viewModel
        userViewModel = ViewModelProvider(this).get(UserViewModel::class.java) // instance
        binding.user = userViewModel // activity_main.xml의 data쪽에서 호출된 name이다.

        // 위에 주석 달은 걸 줄이면 이거임. 데이터를 가져옴.
        // 동작을 받는 역할로 추정
        userViewModel.height.observe(this, Observer {
            binding.textViewHeight.text = it.toString()

            //log
            userViewModel.dataLog()
            Log.i(TAG, "onCreate 후 viewModel 작동함.")
        })

        // ViewModel을 기법으로 작동되는 원리
        // 동작을 주는 역할로 추정
        binding.edt.addTextChangedListener(object : TextWatcher {

            override fun beforeTextChanged(s: CharSequence?, start: Int, count: Int, after: Int) {
            }

            override fun onTextChanged(s: CharSequence?, start: Int, before: Int, count: Int) {
                userViewModel.setData(count) // 보아하니 set은 따로 만들어줘야 하나 보다.

                userViewModel.dataLog()
            }

            override fun afterTextChanged(s: Editable?) {
            }
        })
    }

    override fun onDestroy() {
        Log.i(TAG, "Destroyed")
        super.onDestroy()
    }
}
// 원리 정리
// 1.xml에서 onClick, OnTextChanged OnMethod에서 동작을 판별한다.
// 2.동작을 구분하면, 내가 지정해둔 android:onClick="@{()->user.height}"이라는 구조로 viewModel의 method로 이동한다.
// 3.viewModel의 height는 값이 변하지 않는 상태로 _height -> height로 값을 가져온다.
// 4.height의 변수가 작동되는것을 확인하는 것은 MainActivity의 height.observe다.
// 5.set은 다른 원리로 접근을 해야 한다.